src/app/campaign/store/effects/campaign.effects.ts

```diff
@@ -25,34 +25,124 @@ export class CampaignEffects {
+          ({
+            campaignDoc,
+            updatedFlightDocs,
+            deletedFlightDocs
+          }: {
+            campaignDoc: HalDoc;
+            updatedFlightDocs: { [id: number]: HalDoc };
+            deletedFlightDocs: { [id: number]: HalDoc };
+          }) => {
+            const createdFlights: { [id: number]: Observable<HalDoc> } = payload.createdFlights.reduce(
+              (acc, flight) => ({ ...acc, [flight.id]: this.campaignService.createFlight(flight) }),
+              {}
+            );
+            return payload.createdFlights.length
+              ? forkJoin(createdFlights).pipe(
+                  map(createdFlightDocs => ({ campaignDoc, updatedFlightDocs, createdFlightDocs, deletedFlightDocs }))
+                )
+              : of({ campaignDoc, updatedFlightDocs, deletedFlightDocs });
+          }
+        ),
+        tap(
+          ({
+            campaignDoc,
+            deletedFlightDocs,
+            updatedFlightDocs,
+            createdFlightDocs
+          }: {
+            campaignDoc: HalDoc;
+            deletedFlightDocs: { [id: number]: HalDoc };
+            updatedFlightDocs: { [id: number]: HalDoc };
+            createdFlightDocs: { [id: number]: HalDoc };
+          }) => {
+            this.toastr.success('Campaign saved');
+            // navigate to newly created campaign/flight or away from deleted flight
+            const campaignId = campaignDoc.id;
+            const flightId = +this.router.url.split('/flight/').pop();
+            if (this.router.url.includes('/flight/') && deletedFlightDocs && deletedFlightDocs[flightId]) {
+              // routed flight deleted
+              this.router.navigate(['/campaign', campaignId]);
+            } else if (this.router.url.includes('/flight/') && createdFlightDocs && createdFlightDocs[flightId]) {
+              // flight id changed (flight created)
+              this.router.navigate(['/campaign', campaignId, 'flight', createdFlightDocs[flightId].id]);
+            } else if (!payload.campaign.id || payload.campaign.id !== campaignId) {
+              // campaign id changed (campaign created)
+              this.router.navigate(['/campaign', campaignId]);
+            }
+          }
+        ),
+        map(
+          ({
+            campaignDoc,
+            deletedFlightDocs,
+            updatedFlightDocs,
+            createdFlightDocs
+          }: {
+            campaignDoc: HalDoc;
+            deletedFlightDocs: { [id: number]: HalDoc };
+            updatedFlightDocs: { [id: number]: HalDoc };
+            createdFlightDocs: { [id: number]: HalDoc };
+          }) => new actions.CampaignSaveSuccess({ campaignDoc, deletedFlightDocs, updatedFlightDocs, createdFlightDocs })
+        ),
+        catchError(error => of(new actions.CampaignSaveFailure({ error })))
       );
     })
   );
 
-  constructor(private actions$: Actions, private campaignService: CampaignService, private router: Router) {}
+  @Effect()
```

Comment:
I used effects for add/dup flight because it generates the temporary id (nondeterministic) and routes to the new flight with the id.